runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,224 @@
+            } catch (Throwable throwable) {
+                log.warn("{}:onError at seek to {}", listener.toString(), startAddress + 1);
+                listener.onError(throwable);
+                streamManager.unsubscribe(listener);
+                return;
+            }
+
+            try {
+                while (!shutdown.get()) {
+                    while (!txnStream.hasNext()) {
+                        log.trace("{}::run() all caught up. Taking a 50ms nap", listener.toString());
+                        final int defaultTimeoutInMilli = 50;
+                        TimeUnit.MILLISECONDS.sleep(defaultTimeoutInMilli);
+                    }
+                    ILogData logData = txnStream.nextUpTo(Address.MAX);
+                    MultiObjectSMREntry multiObjSMREntry = (MultiObjectSMREntry) logData.getPayload(runtime);
+                    long epoch = logData.getEpoch();
+                    Map<TableSchema, List<CorfuStreamEntry>> entries = new HashMap<>();
+                    // first only filter by the stream IDs of interest that are present in this logData
+                    logData.getStreams().stream().filter(tablesOfInterest::containsKey)
+                            .forEach(streamId -> entries.put(tablesOfInterest.get(streamId),
+                                    // Only extract the list of updates per stream as a list
+                                    multiObjSMREntry.getSMRUpdates(streamId).stream().map(smrEntry ->
+                                            CorfuStreamEntry.fromSMREntry(smrEntry,
+                                                    epoch,
+                                                    tablesOfInterest.get(streamId).getKeyClass(),
+                                                    tablesOfInterest.get(streamId).getPayloadClass(),
+                                                    tablesOfInterest.get(streamId).getMetadataClass())
+                                    ).collect(Collectors.toList())));
+
+                    if (!entries.isEmpty()) {
+                        CorfuStreamEntries callbackResult = new CorfuStreamEntries(entries);
+                        log.debug("{}::onNext with {} updates", listener.toString(), entries.size());
+                        long onNextStart = System.nanoTime();
+                        listener.onNext(callbackResult);
+                        long onNextEnd = System.nanoTime();
+                        if (TimeUnit.NANOSECONDS.toSeconds(onNextEnd - onNextStart) > 0) {
+                            log.debug("{}::onNext took {}s", listener.toString(),
+                                    TimeUnit.NANOSECONDS.toSeconds(onNextEnd - onNextStart));
+                        }
+                    }
+                }
+            } catch (Throwable throwable) {
+                log.warn("{}::onError: {}", listener.toString(), throwable.toString());
+                listener.onError(throwable);
+                streamManager.unsubscribe(listener);
```

Comment:
This will eventually call interrupt on its own self. After the interrupt there is a log line. Once a thread is interrupted it's possible that the subsequent io call fails with ClosedByInterruptException or similar interrupted exceptions. 

Anyways, it seems like the unsubscribe was designed to be called from a different thread, not the thread itself.